FreeMonoid.agda

```diff
@@ -0,0 +1,72 @@
+module FreeMonoid where
+
+open import Monoid
+open import Relation.Binary
+open import Relation.Binary.PropositionalEquality
+import Relation.Binary.SetoidReasoning as ≈-Reasoning
+
+
+module Carrier (A : Set) where
+
+  data ExpMonoid : Set where
+    Embed : A → ExpMonoid
+    MEmpty : ExpMonoid
+    _<>_ : ExpMonoid → ExpMonoid → ExpMonoid
+
+
+  data EquivMonoid : ExpMonoid → ExpMonoid → Set where
+    equiv-refl : {e : ExpMonoid}
+               → EquivMonoid e e
+    equiv-sym : {e₁ e₂ : ExpMonoid}
+              → EquivMonoid e₁ e₂ → EquivMonoid e₂ e₁
+    equiv-trans : {e₁ e₂ e₃ : ExpMonoid}
+                → EquivMonoid e₁ e₂
+                → EquivMonoid e₂ e₃
+                → EquivMonoid e₁ e₃
+    equiv-cong : {e₁ e₂ e₁′ e₂′ : ExpMonoid}
+               → EquivMonoid e₁ e₁′
+               → EquivMonoid e₂ e₂′
+               → EquivMonoid (e₁ <> e₂) (e₁′ <> e₂′)
+    left-id-mempty : {e : ExpMonoid}
+                   → EquivMonoid (MEmpty <> e) e
+    right-id-mempty : {e : ExpMonoid}
+                    → EquivMonoid (e <> MEmpty) e
+    assoc-mappend : (e₁ e₂ e₃ : ExpMonoid)
+                  → EquivMonoid ((e₁ <> e₂) <> e₃) (e₁ <> (e₂ <> e₃))
+
+  equiv-setoid : Setoid _ _
+  equiv-setoid = record
+    { Carrier = ExpMonoid
+    ; _≈_ = EquivMonoid
+    ; isEquivalence = record
+      { refl  = equiv-refl
+      ; sym   = equiv-sym
+      ; trans = equiv-trans
+      }
+    }
+
+
+  eqClass : ExpMonoid → List A
+  eqClass = {!!}
+
+
+  module Sound where
+    open ≈-Reasoning
+
+    sound : {e₁ e₂ : ExpMonoid}
+          → eqClass e₁ ≡ eqClass e₂
+          → EquivMonoid e₁ e₂
+    sound {e₁} {e₂} prf =
+      begin⟨ equiv-setoid ⟩
+        e₁
```

Comment:
This proof is surprisingly convoluted! Instead of pattern-matching on `e₁` and `e₂`, I had to define an intermediate transformation and use lemmas about that transformation to complete this proof.

The fact that I had to do this is why I prefer to use my intuitive definition of free structures even though I know that the correct definition is the left-adjoint one. When I wrote the left-adjoint proof, I simply followed the types and focused on the next proof step which was in front of me. I'm glad that this technique was sufficient, because if it wasn't, I would have been stuck and I wouldn't know what else to try. With my intuitive definition, however, once I got stuck on this proof of `sound`, I could take a step back and think about why I thought that the statement was true, and then convert that informal reasoning into a proper formal proof. I can't do that if I don't have an intuitive understanding of what the statement means.